iphone/CoreApi/CoreApi/Logger/Logger.mm

```diff
@@ -1,32 +1,180 @@
+  self = [super init];
+  if (self) {
+    _fileSize = 0;
+    _osLogger = os_log_create(kLoggerCategory.UTF8String, kLoggerSubsystem.UTF8String);
+  }
+  return self;
+}
+
+// MARK: - Public
+
++ (void)setFileLoggingEnabled:(BOOL)fileLoggingEnabled {
+  fileLoggingEnabled ? [self enableFileLogging] : [self disableFileLogging];
+}
+
++ (void)log:(LogLevel)level message:(NSString *)message {
+  LOG_SHORT([self baseLevel:level], (message.UTF8String));
 }
 
 + (BOOL)canLog:(LogLevel)level {
   return [Logger baseLevel:level] >= base::g_LogLevel;
 }
 
++ (nullable NSURL *)getLogFileURL {
+  Logger * logger = [self logger];
+  return [logger isFileLoggingEnabled] ? [NSURL fileURLWithPath:logger.filePath] : nil;
+}
+
++ (uint64_t)getLogFileSize {
+  return [self logger].fileSize;
+}
+
+// MARK: - C++ injection
+
+void LogMessage(base::LogLevel level, base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:level src:src message:message];
+  CHECK_LESS(level, base::g_LogAbortLevel, ("Abort. Log level is too serious", level));
+}
+
+bool AssertMessage(base::SrcPoint const & src, std::string const & message)
+{
+  [Logger logMessageWithLevel:base::LCRITICAL src:src message:message];
+  return true;
+}
+
+// MARK: - Private
+
++ (void)enableFileLogging {
+  Logger * logger = [self logger];
+  NSFileManager * fileManager = [NSFileManager defaultManager];
+
+  // Create a log file if it doesn't exist and setup file handle for writing.
+  NSString * filePath = [[fileManager temporaryDirectory] URLByAppendingPathComponent:kLogFileName].path;
+  if (![fileManager fileExistsAtPath:filePath])
+    [fileManager createFileAtPath:filePath contents:nil attributes:nil];
+
+  NSFileHandle * fileHandle = [NSFileHandle fileHandleForWritingAtPath:filePath];
+  if (fileHandle == nil) {
+    LOG(LERROR, ("Failed to open log file for writing", filePath.UTF8String));
+    [self setFileLoggingEnabled:NO];
+    return;
+  }
+  // Clean up the file if it exceeds the maximum size.
+  if ([fileManager contentsAtPath:filePath].length > kMaxLogFileSize)
+    [fileHandle truncateFileAtOffset:0];
+
+  logger.filePath = filePath;
+  logger.fileHandle = fileHandle;
+  LOG(LINFO, ("File logging is enabled"));
+}
+
++ (void)disableFileLogging {
+  Logger * logger = [self logger];
+
+  [logger.fileHandle closeFile];
+  logger.fileHandle = nil;
+  logger.fileSize = 0;
+  NSError * error;
+  [NSFileManager.defaultManager removeItemAtPath:logger.filePath error:&error];
+  if (error)
+    LOG(LERROR, ("Failed to remove log file", logger.filePath.UTF8String));
+  LOG(LINFO, ("File logging is disabled"));
+}
+
+- (BOOL)isFileLoggingEnabled {
+  return self.fileHandle != nil;
+}
+
++ (void)logMessageWithLevel:(base::LogLevel)level src:(base::SrcPoint const &)src message:(std::string const &)message {
+  // Build the log message string.
+  auto & logHelper = base::LogHelper::Instance();
+  std::ostringstream output;
+  logHelper.WriteProlog(output, level);
```

Comment:
This should be synchronized, because GetThreadID inside is single-threaded.

Reply:
`// TODO(KK): Either guard this call, or refactor thread ids in logHelper.`

Did you check it in debug build when many threads are downloading some maps? Does it work well without synchronization?